### PR TITLE
n-dimensional pMCMC

### LINKED ISSUE #9: Generalise scan to arbitrary parameters

### PR DESCRIPTION
**Pull request by johnlees**
This allows sampling of any parameter in `pars_obs` or `model_params`. The input to the pMCMC function is now a data frame which describes the parameters to sample, giving the init, min, max and whether discrete for each. e.g. for the most recent 9D case:
```
           names         init min   max discrete
1     beta_start 1.400000e-01   0 1e+00    FALSE
2       beta_end 3.332000e-02   0 1e+00    FALSE
3     start_date 1.829900e+04   0 1e+06     TRUE
4   gamma_triage 5.099579e-01   0 1e+00    FALSE
5   gamma_hosp_R 1.092046e-01   0 1e+00    FALSE
6   gamma_hosp_D 2.911154e-01   0 1e+00    FALSE
7    gamma_ICU_R 3.541429e-01   0 1e+00    FALSE
8    gamma_ICU_D 2.913861e-01   0 1e+00    FALSE
9 gamma_stepdown 4.523810e-01   0 1e+00    FALSE
```

You can also give priors to each parameter. I'd appreciate input on how best to do this, at the moment this is a function for each parameter, which takes the set of parameters (in case they depend on each other, which is more likely for the complex ones like beta and date than the gammas etc). 

Would it be better to give a single function rather than one for each parameter? Would a helper function to set this up be useful? At the moment to set the previous prior you need to do this:
```
                  pars_lprior = list('beta_start'     = function(pars) log(1e-10),
                                     'beta_end'       = function(pars) 0,
                                     'start_date'     = function(pars) 0,
                                     'gamma_triage'   = function(pars) 0,
                                     'gamma_hosp_R'   = function(pars) 0,
                                     'gamma_hosp_D'   = function(pars) 0,
                                     'gamma_ICU_R'    = function(pars) 0,
                                     'gamma_ICU_D'    = function(pars) 0,
                                     'gamma_stepdown' = function(pars) 0),
```
which is clunky

Closes #9 (finally!)

### CODE REVIEW

**Comment by richfitz** [comment]
I believe that this will need looking at in light of the changes rushed through in #89 and #90 sorry

File: tests/testthat/test-pmcmc.R [hunk 1]
```diff
@@ -350,63 +280,76 @@ test_that("pmcmc with new model", {
...
+                     'start_date'     = function(pars) 0,
+                     'gamma_triage'   = function(pars) 0,
+                     'gamma_hosp_R'   = function(pars) 0,
+                     'gamma_hosp_D'   = function(pars) 0,
+                     'gamma_ICU_R'    = function(pars) 0,
+                     'gamma_ICU_D'    = function(pars) 0,
+                     'gamma_stepdown' = function(pars) 0)
+  
+  proposal_kernel <- diag(length(pars_to_sample$names)) * 0.01^2
+  row.names(proposal_kernel) <- colnames(proposal_kernel) <- pars_to_sample$names
```
Comment (lwhittles) [inline tests/testthat/test-pmcmc.R:341]: Could we add a test that checks the output is robust to the order in which the parameters are entered into pars_to_sample. I'm thinking specifically wrt the proposal kernel?

**Comment by lwhittles** [comment]
This is really really great John, thanks so much for all the work!

File: R/pmcmc.R [hunk 2]
```diff
@@ -181,37 +182,59 @@ pmcmc <- function(data,
...
+  if(start_date_to_offset(data$date[1], pars_init[['start_date']]) < 0) {
+    stop('start date must not be before first date of supplied data')
+  }
 
   #
   # Generate MCMC parameters
   #
+  
+  # This gets changed to numeric in data.frame
+  pars_init[['start_date']] <- as.Date(pars_init[['start_date']], origin = "1970-01-01")
```
[the remark this answers is not in the document]
Comment (johnlees) [reply R/pmcmc.R:236]: This is more to do with how R stores dates, and undoes the automatic conversion. I'll raise an issue for looking at the date offset more generally

[the remark this answers is not in the document]
Comment (johnlees) [reply R/pmcmc.R:236]: Logged as #108

File: tests/testthat/test-pmcmc.R (hunk 1 above)
Comment (johnlees) [reply tests/testthat/test-pmcmc.R:341]: Great idea, this is indeed a bug in the current version

**Comment by lwhittles** [comment]
Just a couple of tiny things then good to go I reckon!

File: R/parameter_proposals.R [hunk 3]
```diff
@@ -5,23 +5,10 @@ update_beta <- function(sircovid_model,
                         beta_pl,
                         start_date,
                         dt) {
-  if(!is.null(beta_pl)) {
     new_beta <- sircovid_model$generate_beta_func(beta_start = beta_start,
```
[the remark this answers is not in the document]
Comment (johnlees) [reply R/parameter_proposals.R:8]: I think it was just doing the same check in two functions, but good that we're being consistent!

File: R/pmcmc.R [hunk 4]
```diff
@@ -176,14 +176,16 @@ pmcmc <- function(data,
...
   }
   
-  correct_format_mat <- function(pars) {
+  correct_format_mat <- function(pars, names) {
        setequal(rownames(pars),
-                colnames(pars)) & 
+                colnames(pars)) && 
          setequal(rownames(pars),
-                   par_names)
+                   par_names) &&
```
[the remark this answers is not in the document]
Comment (johnlees) [reply R/pmcmc.R:183]: ah yes, I guess it still works but you're definitely correct

**Comment by johnlees** [comment]
Dates have been removed by making everything reference days since 2019-12-31. They are readded when plotting, and there is an automatic conversion of the date column in the data object (in `particle_filter_data`), otherwise every date input must be in this new format, which can be achieved by passing through the `sircovid_date()` helper function

**Comment by lwhittles** [comment]
I have the task running and it all looks good to me, a couple of minor comments, but will deal with them in the next version as we need this for tonight

File: R/parameters.R [hunk 5]
```diff
@@ -357,6 +355,27 @@ generate_parameters_base <- function(
...
+##' 
+##' @title sircovid date
+##' 
+##' @param date Date as a string, or a format otherwise understood by lubridate
+##' 
+##' @return days that \code{date} is after the beginning of 2020
+##' 
+##' @import lubridate
+##' 
+sircovid_date <- function(date) {
```
[the remark this answers is not in the document]
Comment (johnlees) [reply R/parameters.R:371]: I had intended to do this, but must have taken it out at some point!

**Comment by johnlees** [comment]
I will add those changes into #111

**Pull request merged**